numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -2236,7 +2236,9 @@ VOID_copyswapn (char *dst, npy_intp dstride, char *src, npy_intp sstride,
         ((PyArrayObject_fields *)arr)->descr = descr;
         return;
     }
-    if (swap && PyArray_DESCR(arr)->subarray != NULL) {
+    if (PyDataType_REFCHK(PyArray_DESCR(arr)) ||
+                /* Cannot have objects without fields or subarrays */
+                (swap && PyArray_DESCR(arr)->subarray != NULL)) {
```

Comment:
This is wrong - the below refers to `descr->subarray`, which previously guaranteed not to be null. Why is a reference check needed here?

Reply:
I think perhaps you want `(PyDataType_REFCHK(PyArray_DESCR(arr)) || swap) && PyArray_DESCR(arr)->subarray != NULL)`.

It's not clear to me that the shortcut for `swap` makes sense in the first place, and it seems it just screws over user-defined types.

Reply:
Oh, moved the checks around. It is correct because of the early return in the previous if, but it reads strange like this...

Reply:
Correct in the sense of the inital bug. But you are right, that was wrong. Also the subarray can be of VOID type again, so then you, again, need copyswap.

Added both with comments, I think it should be right now.

Reply:
I still think the way I wrote it makes more sense...